src/fluree/db/query/fql/parse.cljc

```diff
@@ -0,0 +1,428 @@
+                           (assoc ::where/datatype const/$xsd:anyURI))
+                triple* [s p o*]]
+            (where/->pattern :iri triple*))
+          triple)))))
+
+(defmethod parse-pattern :triple
+  [triple _ db context]
+  (parse-triple triple db context))
+
+(defmethod parse-pattern :union
+  [{:keys [union]} vars db context]
+  (let [parsed (mapv (fn [clause]
+                       (parse-where-clause clause vars db context))
+                     union)]
+    (where/->pattern :union parsed)))
+
+(defmethod parse-pattern :optional
+  [{:keys [optional]} vars db context]
+  (let [clause (if (coll? (first optional))
+                 optional
+                 [optional])
+        parsed (parse-where-clause clause vars db context)]
+    (where/->pattern :optional parsed)))
+
+(defn parse-where
+  [q vars db context]
+  (when-let [where (:where q)]
+    (parse-where-clause where vars db context)))
+
+(defn parse-selector
+  [db context depth s]
+  (cond
+    (syntax/variable? s) (-> s parse-var-name select/variable-selector)
+    (syntax/query-fn? s) (let [{:keys [variable function]} (parse-aggregate s)]
+                           (select/aggregate-selector variable function))
+    (select-map? s)      (let [{:keys [variable selection depth spec]}
+                               (parse-subselection db context s depth)]
+                           (select/subgraph-selector variable selection depth spec))))
+
+(defn parse-select-clause
+  [clause db context depth]
+  (if (sequential? clause)
+    (mapv (partial parse-selector db context depth)
+          clause)
+    (parse-selector db context depth clause)))
+
+(defn parse-select
+  [q db context]
+  (let [depth      (or (:depth q) 0)
+        select-key (some (fn [k]
+                           (and (contains? q k) k))
+                         [:select :selectOne :select-one])
+        select     (-> q
+                       (get select-key)
+                       (parse-select-clause db context depth))]
+    (case select-key
+      :select     (assoc q :select select)
+      :select-one (assoc q :select-one select)
+      :selectOne  (-> q
+                      (dissoc :selectOne)
+                      (assoc :select-one select)))))
+
+(defn ensure-vector
+  [x]
+  (if (vector? x)
+    x
+    [x]))
+
+(defn parse-grouping
+  [q]
+  (some->> (or (:groupBy q)
+               (:group-by q))
+           ensure-vector
+           (mapv parse-var-name)))
+
+(defn parse-ordering
+  [q]
+  (some->> (or (:order-by q)
+               (:orderBy q))
+           ensure-vector
+           (mapv (fn [ord]
+                   (if-let [v (parse-var-name ord)]
+                     [v :asc]
+                     (let [[dir dim] ord
+                           v         (parse-var-name dim)]
+                       (if (syntax/asc? dir)
+                         [v :asc]
+                         [v :desc])))))))
+
+(defn parse-analytical-query
+  [q db]
+  (syntax/validate q)
```

Comment:
I've never used spec as a validation step in production before, does/could this add any meaningful performance overhead?

Reply:
Yes, it could, but I'm not concerned at this point. There is definitely room for optimization in the current implementation as we throw away a lot of the work spec has done to decompose the query into components during validation only to go back and do the same decomposition during parsing. But, beyond that, I think we can gain a lot from normalizing and caching queries more than we do. If, after all that, spec still proves to be the bottleneck in query performance, and that proves to be a problem, it isn't that hard to pull out this validation path and go back to the old way. I think we can cross that bridge when we get to it.

Reply:
Probably worth coming to a decision on it however before writing code to turn s/explain output into good error messages. While it might have changed in the years since introduction, when spec was originally release it was recommended to not use it as an input validation and that was not a use case they were going to focus on supporting - which does introduce risk that future spec changes could further negatively impact this use case.

Reply:
Malli might be an alternative here - it's had a stable release and it's designed to be fast.

Reply:
Yeah, that's probably a better bet. @bplatz makes a good point about the aims of spec. Even if it's fast enough for us now, the fact that production performance isn't a goal of their development should still give us a lot of pause. Malli offers the same value prop and their development goals align more closely with our needs.

Reply:
I second (third?) that, Malli would probably be a great choice here.

Thanks to having all the specs in one place, it shouldn't be too painful to change.